contrib/mixin/mixin.libsonnet

```diff
@@ -227,14 +227,14 @@
           {
             alert: 'etcdExcessiveDatabaseGrowth',
             expr: |||
-              increase(((etcd_mvcc_db_total_size_in_bytes/etcd_server_quota_backend_bytes)*100)[240m:1m]) > 50
+              predict_linear((etcd_mvcc_db_total_size_in_bytes)[4h:], 4*60*60) > etcd_server_quota_backend_bytes
```

Comment:
I am not quite confident about this. Shouldn't it be something like below? 
```
predict_linear(etcd_mvcc_db_total_size_in_bytes[4h], 4*60*60) > etcd_server_quota_backend_bytes
```

Reply:
```
max(predict_linear(etcd_mvcc_db_total_size_in_bytes{%(etcd_selector)s}[4h], 4*60*60)) > etcd_server_quota_backend_bytes{%(etcd_selector)s}
```

looks lovely to me.  I've attached a screenshot of some relevant queries demonstrating this approach in a random cluster.

![image](https://user-images.githubusercontent.com/209920/177925237-3d7a5ac4-2608-41dc-be70-d6418efeb71b.png)

Part of me is now wondering if it's worth using the linear prediction with some kind of averaged current value to guess at when the linear prediction thinks we'll hit the threshold (4h?  2h?  15m?).  But the rest of me is trying to squash that idea with "look at that screenshot!  Clearly the linear prediction is bananas.  4h from 6:30 PM, we are not going to be down around ~500 MB".  So +1 from my to sticking with "Predicting running out of disk space in the next four hours..." and not trying to get more squeeze out more detail than it can reliably deliver :)

![image](https://user-images.githubusercontent.com/209920/177926820-87228d39-48ac-48da-8f10-2c14a3edbed3.png)

Reply:
Although, hmm.  Looking at the last 4h and then extrapolating 4h into the future is... going to be cutting it pretty fine in a lot of "runaway DB growth" cases.  Also, do we want to try and tease apart "so you should really stop that runaway that's spewing data into your DB" and "so you should compact your database and reclaim all that free space churny workloads have been leaving behind"?  Same cluster as the previous comment, comparing `etcd_mvcc_db_total_size_in_bytes` and the "what we'd have if we were compacted" `etcd_mvcc_db_total_size_in_use_in_bytes` (both scaled to a y axis where 1 is "1 GB").  Also mixing in `etcd_debugging_mvcc_db_compaction_last` (scaled to a y axis where 1 is "1 minute").

![image](https://user-images.githubusercontent.com/209920/177929536-cf689e4f-a260-4602-9068-587a9d49dd18.png)

That's a much shorter time window so we don't under-sample the 5s compaction period.  This is a bit confusing to me:

1. `etcd_mvcc_db_total_size_in_use_in_bytes <= etcd_mvcc_db_total_size_in_bytes`, which makes sense.
2. `etcd_mvcc_db_total_size_in_bytes` has plateaus.  Ok, probably churn as resources get added and removed but we'd already claimed enough space, so we don't need to claim more.
3. `etcd_mvcc_db_total_size_in_bytes` occasionally drops down to match `etcd_mvcc_db_total_size_in_use_in_bytes`.  Ok, this is probably compaction.
4. `time() - etcd_debugging_mvcc_db_compaction_last` drops to zero every 5m for "we just compacted", and slowly rises afterwards, as the previous compaction fades into the past, until the next compaction drops us back to zero, etc.
5. But many compactions do not drop `size_in_bytes` down from its plateau.  Why not?  Do we need a rarer defrag or something to drop us from those plateaus?  I hunted around a bit in `server/storage/mvcc/metrics.go`, but nothing stood out at me as a defrag metric.
7. Also, the `size_in_use` drop I've focused on does not seem to align with the `compaction_last` timestamp.  And it's basically 2.5m away.  Does compaction really take 2.5m to wrap up?

I do see some small `size_in_use` drops around the compaction points.  Maybe the big `size_in_bytes` drop is a defrag or something else.  But tying this this rapid-growth to the plateaus+drops `size_in_bytes` metric seems like it makes the linear prediction less useful/responsive.  On the other hand, using the smoother `size_in_use` metric might undershoot the actual space being used, and risk folks crashing into the size cap because they forgot to do whatever the thing is that makes `size_in_bytes` drop back down...

Reply:
```
(
  etcd_mvcc_db_total_size_in_use_in_bytes{%(etcd_selector)s}
  -
  min_over_time(etcd_mvcc_db_total_size_in_use_in_bytes{%(etcd_selector)s}[4h])
) / etcd_server_quota_backend_bytes{%(etcd_selector)s}
>
0.2
```

would be an approach for "currently using 20% more of the total capacity, compared to recent lows".  That would pick up runaway growth faster than the `deriv` or `predict_linear` would.  But it will also continue to complain for a while after the runaway growth is mitigated, unless beyond removing the growth, the mitigation also returns you to within 20% capacity of your recent min.  Here's a screenshot with a more sensitive 8% threshold, because the cluster I'm looking at is not volatile enough to trip a 20% threshold:

![image](https://user-images.githubusercontent.com/209920/177936395-5bdcff7d-a242-40d4-9e64-dc23cf4440c1.png)

And here's one that's scaled by the fraction of remaining space for "currently using 20% more of the remaining capacity, compared to recent lows", instead of being scaled by the fraction of total space:

```
  (
    etcd_mvcc_db_total_size_in_use_in_bytes{%(etcd_selector)s}
    -
    min_over_time(etcd_mvcc_db_total_size_in_use_in_bytes{%(etcd_selector)s}[4h])
  )
  /
  (
    etcd_server_quota_backend_bytes{%(etcd_selector)s}
    -
    min_over_time(etcd_mvcc_db_total_size_in_use_in_bytes{%(etcd_selector)s}[4h])
  )
  >
  0.2
```

Here's a screenshot with a more sensitive 10% threshold, because the cluster I'm looking at is not volatile enough to trip a 20% threshold:

![image](https://user-images.githubusercontent.com/209920/177937169-dc0d7a97-019f-4837-b7c5-632d218fd0bc.png)

Reply:
In addition to predict_linear(), the node filesystem alerts also look at the used space ratio to avoid being oversensitive.
See https://github.com/prometheus/node_exporter/blob/690efa61e86acefdf05bb4334a3d68128ded49c9/docs/node-mixin/alerts/alerts.libsonnet#L10-L16

Reply:
@simonpasquier  thanks a lot for your review, do you recommend adding this to the alert expr ?

Reply:
> In addition to predict_linear(), the node filesystem alerts also look at the used space ratio to avoid being oversensitive. See https://github.com/prometheus/node_exporter/blob/690efa61e86acefdf05bb4334a3d68128ded49c9/docs/node-mixin/alerts/alerts.libsonnet#L10-L16

It makes sense, but I suggest to address it in a separate PR.